### ISSUE TITLE
`groupby` does not correctly handle non-dimensional coordinate

### ISSUE THREAD

**Issue by fujiisoup** [pydata/xarray#2944]
#### Code Sample, a copy-pastable example if possible

```python
>>> import numpy as np
>>> import xarray as xr
>>> 
>>> da = xr.DataArray(np.arange(12).reshape(3, 4), dims=['x', 'y'],
...                   coords={'x': [0, 1, 1], 'x2': ('x', ['a', 'b', 'c'])})
>>> grouped = da.groupby('x').mean('x')
>>> grouped
<xarray.DataArray (x: 2, y: 4)>
array([[0., 1., 2., 3.],
       [6., 7., 8., 9.]])
Coordinates:
  * x        (x) int64 0 1
    x2       (x) <U1 'a' 'b' 'c'  # <-- this has length 3!
Dimensions without coordinates: y
```

#### Problem description
`groupby` operation ignores the non-dimensional coordinate.

#### Expected Output
```python
<xarray.DataArray (x: 2, y: 4)>
array([[0., 1., 2., 3.],
       [6., 7., 8., 9.]])
Coordinates:
  * x        (x) int64 0 1
Dimensions without coordinates: y
```
Compute mean of the coordinate. If not possible, drop it.

**Comment by dcherian** [comment]
So that has been fixed on master somehow but

``` python
da = xr.DataArray(np.arange(12).reshape(3, 4), dims=['x', 'y'],
                  coords={'x': [0, 1, 1], 
                          'x2': ('x', ['a', 'b', 'c']),
                          'x3': ('x', [4, 5, 6])})
da.groupby("x").mean("x")
```
gives
```
<xarray.DataArray (x: 2, y: 4)>
array([[0., 1., 2., 3.],
       [6., 7., 8., 9.]])
Coordinates:
  * x        (x) int64 0 1
Dimensions without coordinates: y
```

i.e coordinate `x3` has been dropped though the mean along `x` is meaningful. This is consistent with the behaviour of `reduce` e.g.
``` python
da.mean("x")

<xarray.DataArray (y: 4)>
array([4., 5., 6., 7.])
Dimensions without coordinates: y
```

What would be the best way to implement reductions for non-dim coords? Add a `reduce_coords` kwarg?

**Comment by max-sixty** [comment]
I had generally thought that we didn't want to keep non-dim coords along the group by i.e. if you want to keep them, put them as variables.

I'm also think it would be difficult to resolve issues like whether or not to keep coords based on their type in a way that wasn't surprising (though I guess we have to do it for data vars, even if we don't hit it often).

**Comment by fujiisoup** [comment]
Closed as the discussions can be continued in #5361 .
